perfkitbenchmarker/providers/gcp/gcp_dpb_dataflow.py

```diff
@@ -152,4 +186,303 @@ def GetMetadata(self):
+        distributions[metric['name']['name']] = metric['distribution']
+      else:
+        logging.warn(f'Unfamiliar metric type found: {metric}')
+
+    self.job_metrics = {
+      METRIC_TYPE_COUNTER: counters,
+      METRIC_TYPE_DISTRIBUTION: distributions
+    }
+
+  def GetMetricValue(self, name, type=METRIC_TYPE_COUNTER):
+    """Get value of a job's metric.
+    
+    Returns:
+      Integer if metric is of type counter
+      Dictionary if metric is of type distribution. Dictionary 
+      contains keys such as count/max/mean/min/sum
+    """
+    if type not in (METRIC_TYPE_COUNTER, METRIC_TYPE_DISTRIBUTION):
+      raise ValueError(f'Invalid type provided to GetMetricValue(): {type}')
+    
+    if self.job_metrics is None:
+      self._PullJobMetrics()
+    
+    return self.job_metrics[type][name]
+
+  def GetAvgCpuUtilization(self, start_time: datetime, end_time: datetime):
+    """Get average cpu utilization across all pipeline workers.
+
+    Args:
+      start_time: datetime specifying the beginning of the time interval.
+      end_time: datetime specifying the end of the time interval.
+
+    Returns:
+      Average value across time interval
+    """
+    client = monitoring_v3.MetricServiceClient()
+    project_name = f"projects/{self.project}"
+    
+    now_seconds = int(time.time())
+    start_time_seconds = int(start_time.timestamp())
+    end_time_seconds = int(end_time.timestamp())
+    # Cpu metrics data can take up to 240 seconds to appear
+    if (now_seconds - end_time_seconds) < CPU_API_DELAY_SECONDS:
+      logging.info('Waiting for CPU metrics to be available (up to 4 minutes)...')
+      time.sleep(CPU_API_DELAY_SECONDS - (now_seconds - end_time_seconds))
+
+    interval = TimeInterval(
+        {
+            "start_time": {"seconds": start_time_seconds},
+            "end_time": {"seconds": end_time_seconds},
+        }
+    )
+
+    api_filter = (
+      'metric.type = "compute.googleapis.com/instance/cpu/utilization" '
+      f'AND resource.labels.project_id = "{self.project}" '
+      f'AND metadata.user_labels.dataflow_job_id = "{self.job_id}" '
+    )
+
+    aggregation = Aggregation(
+        {
+            "alignment_period": {"seconds": 60},  # 1 minute
+            "per_series_aligner": Aggregation.Aligner.ALIGN_MEAN,
+            "cross_series_reducer": Aggregation.Reducer.REDUCE_MEAN,
+            "group_by_fields": ["resource.instance_id"],
+        }
+    )
+
+    results = client.list_time_series(
+      request={
+        "name": project_name,
+        "filter": api_filter,
+        "interval": interval,
+        "view": monitoring_v3.ListTimeSeriesRequest.TimeSeriesView.FULL,
+        "aggregation": aggregation,
+      }
+    )
+
+    # if not results:
+    #   raise Exception('No monitoring data found. Unable to calculate avg CPU utilization')
+    return self._GetAvgValueFromTimeSeries(results)
+
+  def GetMaxOutputThroughput(self, ptransform: str, start_time: datetime, end_time: datetime):
```

Comment:
Does this actually get used in this class? If not I suggest moving it to the subclass that will be in the subsequent PR.